tests/components/zeroconf/test_init.py

```diff
@@ -349,6 +367,39 @@ def http_only_service_update_mock(ipv6, zeroconf, services, handlers):
     assert mock_config_flow.mock_calls[0][1][0] == "samsungtv"
 
 
+async def test_zeroconf_match_am(hass, mock_async_zeroconf):
```

Comment:
Seems like you could use `@pytest.mark.parametrize` for the different properties to avoid having duplicate tests?

Reply:
I didn't do that because it would have to have 6 params which got a bit messy so I figured it was better to write it out.

Reply:
I guess it's a bit tricky, but if most non-relevant values (like service names and platform) are hardcoded to the same things in the test, then you would only need zeroconf properties and matchers like this:

```python
@pytest.mark.parametrize("properties,matchers",
    [
        ({"am": "AudioAccessory5,1"}, {"am": "audioaccessory*"}),
    ]
)
async def test_zeroconf_match_am(hass, mock_async_zeroconf, properties, matchers):
    """Test configured options for a device are loaded via config entry."""

    def http_only_service_update_mock(ipv6, zeroconf, services, handlers):
        """Call service update handler."""
        handlers[0](
            zeroconf,
            "_raop._tcp.local.",
            "XXXXXXXXXXXX@Outside Speakers._raop._tcp.local.",
            ServiceStateChange.Added,
        )

    with patch.dict(
        zc_gen.ZEROCONF,
        {"_raop._tcp.local.": [{"domain": "apple_tv", **matchers}]},
        clear=True,
    ), patch.object(
        hass.config_entries.flow, "async_init"
    ) as mock_config_flow, patch.object(
        zeroconf, "HaAsyncServiceBrowser", side_effect=http_only_service_update_mock
    ) as mock_service_browser, patch(
        "homeassistant.components.zeroconf.AsyncServiceInfo",
        side_effect=get_zeroconf_info_mock(properties),  # Change to include all properties
    ):
        assert await async_setup_component(hass, zeroconf.DOMAIN, {zeroconf.DOMAIN: {}})
        hass.bus.async_fire(EVENT_HOMEASSISTANT_STARTED)
        await hass.async_block_till_done()

    assert len(mock_service_browser.mock_calls) == 1
    assert len(mock_config_flow.mock_calls) == 1
    assert mock_config_flow.mock_calls[0][1][0] == "apple_tv"
```
It would not deal with "not present" cases, so that would have to be a separate test case (probably also parameterized).

Reply:
The downside for me is we loose the context that `am` is really about `_raop._tcp.local.`, and what an `_raop` host is supposed to look like.  That could make it harder for someone trying to maintain this in the future if they make changes since the code is no longer showing the original intent.

Reply:
I'm not sure if I'd remember why it works this way or why we care about matching on `am` 2 years from now without the additional context

Reply:
I agree with that, sure, but my counter question is: why not support matching arbitrary Zeroconf fields instead of picking a few ones that "happens to be of interest"? There's no intent shown in code (only in tests) nor any restrictions as you can match this property with any service. This would have been better IMHO:

```
{"type":"_airplay._tcp.local.", "conditions": {"model":"appletv*"}},
```

Or similar. You get the idea.

Reply:
That would be much better.  I'd have to break the existing matching or add support for yet another way to do it though since we have

```
# Keys we support matching against in properties that are always matched in
# upper case. ex: ZeroconfServiceInfo.properties["macaddress"]
UPPER_MATCH_PROPS = {"macaddress"}
# Keys we support matching against in properties that are always matched in
# lower case. ex: ZeroconfServiceInfo.properties["model"]
LOWER_MATCH_PROPS = {"am", "manufacturer", "model"}
# Top level keys we support matching against in properties that are always matched in
# lower case. ex: ZeroconfServiceInfo.name
```

Reply:
I'll mark this as a draft while I think about what refactoring all the zeroconf matching looks like since it will impact all the integrations using zeroconf

Reply:
We probably want to allow a `properties` `dict`

Reply:
A simple but backwards compatible solution would be to extract all "other" keys in the dict and treat those as matchers. So in this case:

    {"type":"_airplay._tcp.local.", "model":"appletv*", "macaddr": "aa:bb:cc:dd:ee:ff", "foo": "bar"}

We would "pop" `type` and use the rest as matchers. Obviously, this would impose restrictions since you can't match on reserved names (`type` being one, not sure if there are others?). But it's very clean and backwards compatible as mentioned.

Reply:
Its more about rewriting the dev docs and updating all the manifests to the new method 🥲 

I've had 35 hours of flying and 50 hours of travel this week so I'm too wiped out to get it done this week.